### ISSUE TITLE
launch.json: args field is ignored in msys

### ISSUE THREAD

**Issue by alanlivio** [microsoft/vscode-cmake-tools#1630]
### Brief Issue Summary

I using msys2 on windows. 
I configure my launch.json as described [here](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json).
After configure the cmake normaly, I change de the target by the cmake-tools bar and the ${command:cmake.launchTargetPath} is corrected change when launch/debug.
However, the args field at launch.json is ignored.

### Expected:

Launch the ${command:cmake.launchTargetPath} passing the value of launch.json "args" field as param.

### Platform and Versions

- **Operating System**: Windows
- **CMake Version**: 3.19.3
- **VSCode Version**: 1.52.1
- **CMake Tools Extension Version**: v1.5.3
- **Compiler/Toolchain**: gcc.exe (Rev6, Built by MSYS2 project) 10.2.0

**Comment by andreeis** [comment]
If you debug/launch via the CMake Tools commands (cmake.debugTarget and cmake.launchTarget) which are accessibly in the status bar or in the context menus of the CMake Tools left side panel, then nothing is read from launch.json. We have a workitem to unify launch.json with our internal 'on the fly' small launch.json that is created for this feature but we won't get there any time soon.

We have this PR https://github.com/microsoft/vscode-cmake-tools/pull/1603 that attempts to address the use of arguments defined via settings cmake.debugConfig (different than launch.json) but there are some challenges here before we finalize reviewing the PR (different kinds of terminals that can be invoked, quoting scenarios, etc ...).

If you debug/launch via the commands of the Debugger extension, then your launch.json should be read. Note that the default key binding between Debug: "Run without debugging" and "CMake: Debug" overlap, so invoke the the "Debug: Start without debugging" command from the palette and not via keys.

**Comment by alanlivio** [comment]
Hi @andreeis . Thank you so much for your feedback.

My vscode command patella list the following two.
![image](https://user-images.githubusercontent.com/604808/105933548-66d59400-602d-11eb-827c-73adfd776603.png)
"Cmake: Start Without Debuging": build the program with cmake and launch but without read the args in launch.json, as you just mentioned.
"Run: Start  Without Debuging": just build the program with cmake but do not launch the program.

I appreciate in advance your feadback again.

**Comment by andreeis** [comment]
You're saying that "Run: Start without debugging" is not launching anything? Look in the "DEBUG CONSOLE" or "TERMINAL" (near the output channels).

**Comment by alanlivio** [comment]
Hi @andreeis. Thank for your feedback.
Yes. The "Run: Start without debugging" is not launching anything,
When "Cmake: Start Without Debuging", the project is builded and launched.

When  "Run: Start without debugging", the project is only builded
The OUTPUT tab builds the project.
```
[main] Building folder: ginga gingagui
[build] Starting build
[proc] Executing command: C:\tools\msys64\mingw64\bin\cmake.exe --build c:/Users/alan/dev/ginga/ginga/cmake-build-Debug-Windows_NT --config Debug --target gingagui -- -j 3
[build] ninja: no work to do.
[build] Build finished with exit code 0`
```

This is my launch.json based in the [template ](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json).
`````json
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "(gdb) Launch",
      "type": "cppdbg",
      "request": "launch",
      // Resolved by CMake Tools:
      "program": "${command:cmake.launchTargetPath}",
      "args": [
        "/Users/alan/dev/ginga/ginga/examples/primeiro-joao/00syncProp.ncl"
      ],
      "stopAtEntry": false,
      "cwd": "${workspaceFolder}",
      "environment": [
        {
          // add the directory where our target was built to the PATHs
          // it gets resolved by CMake Tools:
          "name": "PATH",
          "value": "$PATH:${command:cmake.launchTargetDirectory}"
        },
        {
          "name": "OTHER_VALUE",
          "value": "Something something"
        }
      ],
      "externalConsole": true,
      "MIMode": "gdb",
      "setupCommands": [
        {
          "description": "Enable pretty-printing for gdb",
          "text": "-enable-pretty-printing",
          "ignoreFailures": true
        }
      ]
    }
  ]
}
````

**Comment by andreeis** [comment]
I reproduce something sometimes. Please share what happens in the terminal after you run the "Run: Start without debugging" command.

**Comment by andreeis** [comment]
It looks that this is a duplicate of https://github.com/microsoft/vscode-cpptools/issues/5680 and it happens when there is an miEngine defined (which you need for your MSYS scenario). There is a workaround described at the end of that report.

**Comment by alanlivio** [comment]
Thanks for the feedback @andreeis.

**Comment by alanlivio** [comment]
@andreeis maybe I was not clear in my question. I have two problems in fact.
1) "Run: Start without debugging" is not working. Your feedback answer that by citing the microsoft/vscode-cpptools#5680 
2) The launch.json: args field is ignored in msys.  In other words, the "Debug: Start Debugging" is not loaded the launch.json args as described [here](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html). This is the main reason of the issue.
Please may you reopen the issue?

**Comment by alanlivio** [comment]
For the 2), the only log I receive in the DEBUG CONSOLE is:  
ERROR: Unable to start debugging. Unexpected GDB output from command "-exec-run". During startup program exited with code 0xc0000135.

**Comment by andreeis** [comment]
Args in launch.json: it usually works. In your case I'm still not sure it doesn't work, you just hit something else before it has a chance. The "-exec-run" gdb error I never saw, I don' know what to advise now, let me dig more about it. Maybe is linked to that other problem with "Run without debugging" which was also a VSCode debugger issue with MSYS (not CMake Tools).

But when you run "CMake: Debug target", that will not read your launch.json. You still can define args for this scenario.
In .vscode/settings.json:
"cmake.debugConfig": {
    "args": ["arg1", "arg2", "arg3"]
}

Let us know if cmake.debugConfig.args unblocks you when you use the CMake Tools debugging instead of the general VSCode debugging.

**Comment by andreeis** [comment]
For the "-exec-run" problem, here is a link. If you don't find anything similar in there to help you we'll look closer.
https://stackoverflow.com/questions/60923232/error-unable-to-start-debugging-unexpected-gdb-output-from-command-exec-run

**Comment by alanlivio** [comment]
Hi @andreeis .  Thanks again for your attention with me.

- Cmake: Start Debugging
Your tip about "cmake.debugConfig": {"args": ["arg1", "arg2", "arg3"]} make it works! Thanks a lot. =) 

- CMake: Start Without Debugging
It still just run the application but does not load the args from the launch.json (as described [here](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json)). Is there a "cmake.runConfig"?
   
- Start Without Debugging
About this. Sadly, I tried the "Unexpected GDB output from command "-exec-run" mentioned in the [link that you gave](https://stackoverflow.com/questions/60923232/error-unable-to-start-debugging-unexpected-gdb-output-from-command-exec-run) but the problem persists. =(
  
- Start Debugging
It still just run the application but does not load the args from the launch.json (as described [here](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json)).

**Comment by alanlivio** [comment]
Hi again. 
I was able to fix  "Unexpected GDB output from command "-exec-run" after remove the "environment" from the launch.json.
Therefore, without using the cmake.debugConfig, I was able to use the "Start Debugging" using the args in from the launch.

This is my used launch.json.

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "cmaketools",
            "type": "cppdbg",
            "request": "launch",
            "program": "${command:cmake.launchTargetPath}",
            "args": [
                "${workspaceFolder}/examples/primeiro-joao/00syncProp.ncl",
            ],
            "cwd": "${workspaceFolder}",
            "setupCommands": [
                {
                    "description": "gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ]
        }
    ]
}
```

I go little further and discover that the problem in fact occour when use of the PATH variable.  
The launch fails If use it with empty value or with the suggested value "$PATH:${command:cmake.launchTargetDirectory}" from [cmakeTools launch.json template](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json)).


This way works.
```json
            "environment": {}                   
```
This way does not work.
```json
            "environment": [
                {
                    "name": "PATH",
                    "value": "$PATH:${command:cmake.launchTargetDirectory}"
                },
```
This way also does not work.
```json
            "environment": [
                {
                    "name": "PATH",
                    "value": "$PATH"
                },
```

Maybe Is there a problem with the [cmakeTools launch.json template](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json))?

Thanks again by the support.

**Comment by andreeis** [comment]
> CMake: Start Without Debugging
> It still just run the application but does not load the args from the launch.json (as described here). Is there a "cmake.runConfig"?
> 

We recently fixed this. Even before we ship the next release, you can try one of our latest vsix-es that the GitHub actions produce after each PR that gets merged in. For example, this one: https://github.com/microsoft/vscode-cmake-tools/actions/runs/665822834#artifacts.


I am investigating the other side of this issue, regarding "environment" in launch.json.

**Comment by andreeis** [comment]
And try to use the VSCode syntax for environment variables ${env:PATH} instead of the linux one $PATH.

**Comment by alanlivio** [comment]
Hi @andreeis . Thank you for the return.

> And try to use the VSCode syntax for environment variables ${env:PATH} instead of the linux one $PATH.  

It worked. Thanks. It should be nice update the [launch.json example] (https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#debugging-launch-json) with this information.

> We recently fixed this. Even before we ship the next release, you can try one of our latest vsix-es that the GitHub actions produce after each PR that gets merged in. For example, this one: https://github.com/microsoft/vscode-cmake-tools/actions/runs/665822834#artifacts.

Nice. But I tried the build and does not work 😞. The output error is show in what next.

```
[cmakefileapi-driver] This version of CMake does not support the "toolchains" object kind. Compiler paths will be determined by reading CMakeCache.txt.
[rollbar] Unhandled exception: Unhandled Promise rejection: Update code model for cpptools TypeError: Cannot read property 'get' of undefined {}
[main] Building folder: ginga-gpmm ginga
[build] Starting build
[proc] Executing command: C:\tools\msys64\mingw64\bin\cmake.exe --build c:/Users/alan/dev/ginga/ginga-gpmm/_build-Debug-Windows_NT --config Debug --target ginga -- -j 6
[build] ninja: no work to do.
[cmakefileapi-driver] This version of CMake does not support the "toolchains" object kind. Compiler paths will be determined by reading CMakeCache.txt.
[build] Build finished with exit code 0
[rollbar] Unhandled exception: Unhandled Promise rejection: Update code model for cpptools TypeError: Cannot read property 'get' of undefined {}
`
```

**Comment by andreeis** [comment]
@alanlivio, indeed we implemented an additional way to calculate compiler paths for a cmake >= 3.20. I missed testing that feature on a system with an older cmake. Of course we don't want to break users and force them to upgrade cmake. If you can, you can upgrade cmake to at least 3.20 and try that vsix again, just to confirm if now all your debugging scenarios work. And remind me if in this thread there is still anything else left un-addressed.
In the meantime, I'll look at what happens with that toolchains object feature on a system with cmake older than 3.20.

**Comment by andreeis** [comment]
We have a fix for the unhandled promise issue that you see. If you wait a few hours, you can install a different vsix to verify how cmake.debugConfig.args works for you.

**Comment by andreeis** [comment]
"CMake: Start without debugging" is also reading cmake.debugConfig.args, there is no need for cmake.runConfig. And I just tried it over here, the args are passed in. I am puzzled why it doesn't work for you.
